repo/ecc/ecc_rs_bw.go

```diff
@@ -0,0 +1,186 @@
+	Options
+	FixedShareSize  bool
+	RequiredShares  int
+	RedundantShares int
+	Fec             *infectious.FEC
+}
+
+func NewRsBwEcc(opts *Options, fixedShareSize bool, spaceUsedPercentage float32) (*RsBwEcc, error) {
+	result := new(RsBwEcc)
+
+	result.Options = *opts
+	result.FixedShareSize = fixedShareSize
+	result.RequiredShares, result.RedundantShares = ComputeShards(spaceUsedPercentage)
+
+	var err error
+	result.Fec, err = infectious.NewFEC(result.RequiredShares, result.RequiredShares+result.RedundantShares)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating infectious FEC")
+	}
+
+	return result, nil
+}
+
+func (r *RsBwEcc) ComputeSizesFromOriginal(length int) (blocks, shareSize, originalSize int) {
+	if r.FixedShareSize {
+		shareSize = 1024
+		blocks = CeilInt(length, r.RequiredShares*shareSize)
+	} else {
+		blocks = 16
+		shareSize = CeilInt(length, r.RequiredShares*blocks)
+	}
+	originalSize = length
+	return
+}
+
+func (r *RsBwEcc) ComputeSizesFromStored(length int) (blocks, shareSize, originalSize int) {
+	if r.FixedShareSize {
+		shareSize = 1024
+		blocks = CeilInt(length, (r.RequiredShares+r.RedundantShares)*shareSize)
+	} else {
+		blocks = 16
+		shareSize = CeilInt(length, (r.RequiredShares+r.RedundantShares)*blocks)
+	}
+	originalSize = length - r.RedundantShares*shareSize*blocks
+	return
+}
+
+func (r *RsBwEcc) Encrypt(input gather.Bytes, contentID []byte, output *gather.WriteBuffer) error {
+	blocks, shareSize, originalSize := r.ComputeSizesFromOriginal(input.Length())
+	requiredSize := r.RequiredShares * shareSize
+
+	var inputBuffer gather.WriteBuffer
+	defer inputBuffer.Close()
+
+	inputBytes := inputBuffer.MakeContiguous(requiredSize * blocks)
+
+	copied := input.AppendToSlice(inputBytes[:0])
+
+	if len(copied) < len(inputBytes) {
+		clear(inputBytes[len(copied):])
+	}
+
+	inputPos := 0
+
+	for b := 0; b < blocks; b++ {
+		block := inputBytes[inputPos : inputPos+requiredSize]
+		inputPos += requiredSize
+
+		err := r.Fec.Encode(block, func(share infectious.Share) {
+			if share.Number >= r.RequiredShares {
+				output.Append(share.Data)
+			}
+		})
+		if err != nil {
+			return errors.Wrap(err, "Error computing ECC")
+		}
+	}
+
+	output.Append(inputBytes[:originalSize])
+
+	return nil
+}
+
+func (r *RsBwEcc) Decrypt(input gather.Bytes, contentID []byte, output *gather.WriteBuffer) error {
+	blocks, shareSize, originalSize := r.ComputeSizesFromStored(input.Length())
+	requiredSize := r.RequiredShares * shareSize
+	redundantSize := r.RedundantShares * shareSize
+
+	var inputBuffer gather.WriteBuffer
+	defer inputBuffer.Close()
+
+	inputBytes := inputBuffer.MakeContiguous((requiredSize + redundantSize) * blocks)
+
+	copied := input.AppendToSlice(inputBytes[:0])
+
+	if len(copied) < len(inputBytes) {
+		clear(inputBytes[len(copied):])
+	}
+
+	eccBytes := inputBytes[:redundantSize*blocks]
+	dataBytes := inputBytes[redundantSize*blocks:]
+
+	shares := make([]infectious.Share, r.RequiredShares+r.RedundantShares)
```

Comment:
Sometimes (not always, though), it may make sense to use arrays on the stack (if you know max capacity), instead of `make()`. This can avoid memory allocations.

I'm guessing the reason for poor throughput for small files is memory allocations:
You can memory allocations using kopia quickly by using:

```
$ kopia <some-subcommand-and-flags> --profile-memory 1 --profile-dir /tmp/profile1
$ go tool pprof -http :9000 -alloc_objects -lines /tmp/profile1/mem.pprof
```

There are similar flags for profiling CPU, mutexes and a bunch more. Golang is awesome.

Reply:
That is nice to know, thanks. I had not realized that `make` allocates in the heap and just creating the array allocates in the stack. I'll play with it.